### PR TITLE
Preliminary HSP support

### PR DESCRIPTION
**Pull request by endrift**
Repurposed for just HSP support for now, since merging that will be easier and I can get to GB Player when this interface is merged.

### CODE REVIEW

File: Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp
Comment (jordan-woyak) [inline Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp]: We have a `MathUtil::NextPowerOf2` that does this.

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp
Comment (jordan-woyak) [inline Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp]: These redundant magic numbers could use a named constant.

**Review by jordan-woyak — DISMISSED** [review]

File: CMakeLists.txt
Comment (lioncash) [inline CMakeLists.txt]: Please don't. Use `target_include_directories` where necessary on a target.

File: CMakeLists.txt
Comment (lioncash) [inline CMakeLists.txt]: You may want to consider making these qualified upstream in mGBA then modifying them here, (like `MGBA_DISASBLE_DEPS`), etc. These variables can now potentially leak into other dependencies. The build is essentially more fragile now.

File: Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp]: This is undefined behavior.

```cpp
u64 value;
std::memcpy(&value, &m_ptr[address & mask], sizeof(value));
return Common::swap64(value);
```

File: Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp]: Ditto.

File: Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.h]:
```suggestion
  explicit CHSPDevice_ARAMExpansion(HSPDevices device);
```

File: Source/Core/Core/HW/HSP/HSP_Device.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_Device.h]:
```suggestion
  explicit IHSPDevice(HSPDevices device_type);
```

File: Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.h]:
```suggestion
  void Write(u32 address, u64 value) override;
```

File: Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.h]:
```suggestion
  ~CHSPDevice_ARAMExpansion() override;
```

File: Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.h]:
```suggestion
  u64 Read(u32 address) override;
```

File: Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.h]:
```suggestion
  void DoState(PointerWrap&) override;
```

File: Source/Core/Core/HW/HSP/HSP_DeviceNull.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceNull.h]:
```suggestion
  explicit CHSPDevice_Null(HSPDevices device);
```

File: Source/Core/Core/HW/HSP/HSP_DeviceNull.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceNull.h]:
```suggestion
  void Write(u32 address, u64 value) override;
```

File: Source/Core/Core/HW/HSP/HSP_DeviceNull.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceNull.h]:
```suggestion
  u64 Read(u32 address) override;
```

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.h]:
```suggestion
  explicit CHSPDevice_GBPlayer(HSPDevices device);
```

The destructor should also be defaulted in the cpp file, since the class contains a non-trivial member like a std::unique_ptr. This allows forward-declaring the class, should it be necessary, without unexpected/confusing compilation errors.

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.h]: override instead of virtual (ditto below)

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.h]:
```suggestion
  explicit IGBPlayer(CHSPDevice_GBPlayer*);
```

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.h
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.h]: These arrays should likely be using `std::array`

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp]: This is undefined behavior.

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp]: This is undefined behavior.

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp]: This can be removed by just adding a `= {};` to the declaration in the class body.

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp]: These constants can be applied directly in the class body.

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp]:
```suggestion
  explicit CGBPlayer_mGBA(CHSPDevice_GBPlayer*);
```

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp]: Override instead of virtual (ditto below)

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp
Comment (lioncash) [inline Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.cpp]: mgba includes should be using `<>` instead of `""`, since it's a dependency.

File: CMakeLists.txt
Comment (lioncash) [reply CMakeLists.txt]: Should have added this during the initial review, but, this can be a lot to ask for a separate project. I mean, going, "lol just change ur build option flag names", isn't realistic in some scenarios, especially if the emulator is already being used as a lib elsewhere, since that'd be a minor forward-breaking change. Because of that, I definitely won't block on this if that's too much of a chaotic change.

Pinging the CMake wiz @Orphis. Is there a nicer way to avoid this that doesn't involve upstream
changes, or is that the only way in this scenario?

File: CMakeLists.txt
Comment (Orphis) [reply CMakeLists.txt]: It would be nicer if the mgba target provided include directories directly.

File: CMakeLists.txt
Comment (Orphis) [reply CMakeLists.txt]: It depends on the extent of functionality that this option adds.
Possibly the best option would be to just use `add_subdirectory(Externals/mgba EXCLUDE_FROM_ALL)` without any special option. The final solution will have all the targets, but they won't be built unless they're actually used.

File: CMakeLists.txt
Comment (Orphis) [inline CMakeLists.txt]: And now, you've changed the option for all the libraries.
You should move this to another file so that it doesn't leak to this project or make the main CMakeLists.txt even longer.

File: CMakeLists.txt
Comment (Orphis) [inline CMakeLists.txt]: Please make sure the `mgba` target exports the right include directories upstream so that we don't need this.

File: CMakeLists.txt
Comment (Orphis) [inline CMakeLists.txt]: Add `EXCLUDE_FROM_ALL`.

File: Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp
Comment (BhaaLseN) [reply Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp]: Or use the handy [`Common::BitCast`](https://github.com/dolphin-emu/dolphin/blob/3b4486bad3f0e44c3a572bba411879a634ece325/Source/Core/Common/BitUtils.h#L189)

**Comment by endrift** [comment]
Will be fixing stuff upstream as needed, don't worry. I'm at lunch right now but I have a good chunk of this stuff resolved already. Will finish when I get home.

**Comment by endrift** [comment]
Might splitting out the base HSP changes from the GBP changes into separate PRs make sense? That was we can get the easy stuff merged first, not that it's actually useful without the GBP changes.

File: Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.h
Comment (endrift) [reply Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.h]: What do you mean by "defaulted in the cpp file"? Is there some C++11 defaulting behavior I'm unfamiliar with?

Comment (lioncash) [reply Source/Core/Core/HW/HSP/HSP_DeviceGBPlayer.h]: e.g. doing the following:

```cpp
// header

class Thing {
public:
    ~CHSPDevice_GBPlayer();
};

// cpp file:

CHSPDevice_GBPlayer::~CHSPDevice_GBPlayer() = default;
```

File: Source/Core/Core/HW/HW.cpp
Comment (BhaaLseN) [inline Source/Core/Core/HW/HW.cpp]: You should increment the [state version](https://github.com/dolphin-emu/dolphin/blob/37c85b32c058e820d0bc81aca8d380e09aef1bd5/Source/Core/Core/State.cpp#L77) as well.

Comment (JosJuice) [reply Source/Core/Core/HW/HW.cpp:103]: This needs fixing again now.

File: Source/Core/Core/HW/HSP/HSP.cpp
Comment (Pokechu22) [inline Source/Core/Core/HW/HSP/HSP.cpp:55]: The call to `RemoveDevice()` is unneeded, as assigning to a `unique_ptr` deletes the old value (if present), as if `reset()` were called.

File: Source/Core/Common/Logging/Log.h
Comment (Pokechu22) [inline Source/Core/Common/Logging/Log.h:37]: These aren't alphabetized properly (the order in LogManager.cpp is correct though).

**Comment by Pokechu22** [comment]
What happens if a DMA starts in-bounds for ARAM, but then goes out of bounds?  Currently I think that would result in writing past the end of `s_ARAM`, but I assume it should instead start writing into HSP?  (i.e. there should be one loop for `while (s_arDMA.Cnt.count)` instead of two based on `if (s_arDMA.ARAddr < s_ARAM.size)`)

File: Source/Core/Core/HW/HSP/HSP.cpp
Comment (Pokechu22) [inline Source/Core/Core/HW/HSP/HSP.cpp:3]: These need to be updated to the new format:

```C++
// Copyright 2018 Dolphin Emulator Project
// SPDX-License-Identifier: GPL-2.0-or-later
```

(you could also bump the year to 2022, though I don't have a strong preference to that)

**Comment by JMC47** [comment]
Once the build issues + review comments are handled, I assume this is ready to merge?

**Comment by endrift** [comment]
I've fixed everything except the ARAM loop and swapping. Unsure how important those are at the moment but I can address them soon if desired.

File: Source/Core/Core/Config/MainSettings.cpp
Comment (shuffle2) [inline Source/Core/Core/Config/MainSettings.cpp:185]: nit: use hex or some multiplies, or some named value, etc...to get rid of the magic base10

Comment (endrift) [reply Source/Core/Core/Config/MainSettings.cpp:185]: Perhaps then the CustomRTCValue default should be changed to 10957 * 24 * 60 * 60 too.

File: Source/Core/Core/HW/HSP/HSP.cpp
Comment (Pokechu22) [inline Source/Core/Core/HW/HSP/HSP.cpp:46]: I don't think this is going to work properly; if `s_device` is not present, then the save state will have a different format from if it is present (although that probably shouldn't happen, since it should only be possible to load states after `Init()` is called and before `Shutdown()` is called).  More importantly, though, this doesn't handle changing devices if the savestate was saved with a different device selected than when it was loaded.

Here's how the EXI version handles this - it's a bit messy, but to my understanding it solves this issue:

https://github.com/dolphin-emu/dolphin/blob/0f2540a0d1133950467845f20b1e003181147781/Source/Core/Core/HW/EXI/EXI_Channel.cpp#L247-L261

(That code also assumes that device is not null, and in fact EXI only checks it if it's directly using the chip select input since some values of that correspond to no device.  The same logic here could remove the checks on whether `s_device` is present, but I don't feel too strongly about whether that should be done.)

**Review by Pokechu22 — COMMENTED** [review]
The swapping isn't too important - it's something that could be addressed in a later PR if it makes things easier there.  The loop issue (https://github.com/dolphin-emu/dolphin/pull/8067#issuecomment-1076904557) is also something that could be addressed in a later PR.

File: Source/Core/Core/Config/MainSettings.cpp
Comment (Pokechu22) [reply Source/Core/Core/Config/MainSettings.cpp:185]: #10684

File: Source/Core/Core/HW/HSP/HSP.cpp
Comment (endrift) [reply Source/Core/Core/HW/HSP/HSP.cpp]: This isn't done in exactly the same way, but I think it's fixed now.

File: Source/Core/DolphinLib.props
Comment (shuffle2) [inline Source/Core/DolphinLib.props:880]: `ClCompile`

Comment (endrift) [reply Source/Core/DolphinLib.props]: I wish I'd gotten a better night's sleep, clearly.

**Comment by JosJuice** [comment]
@dolphin-emu-bot rebuild

File: Source/Core/Core/HW/HSP/HSP.cpp
Comment (Pokechu22) [reply Source/Core/Core/HW/HSP/HSP.cpp]: The new way looks fine to me; EXI's logic is a bit more complicated because it needs to handle hot-swapping (and in some cases making sure games notice that hot-swapping happened by having the device register as via interrupts, but it needs to not do that for a savestate).  I assume it's not normally valid to remove an HSP device while a game is running.

Comment (endrift) [reply Source/Core/Core/HW/HSP/HSP.cpp]: Considering you'd need to pick up the GameCube to remove an HSP device, that seems like a safe bet.

File: Source/Core/Core/HW/HSP/HSP.cpp
Comment (Pokechu22) [inline Source/Core/Core/HW/HSP/HSP.cpp:8]: This include should use quotes, since ChunkFile.h is part of Dolphin.

File: Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp
Comment (Pokechu22) [inline Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp:18]: This include should go at the start (after the include for HSP_DeviceARAMExpansion.h) per [this](https://github.com/dolphin-emu/dolphin/blob/master/Contributing.md#headers).

File: Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp
Comment (Pokechu22) [inline Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp:7]: This include is unused.

File: Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp
Comment (Pokechu22) [inline Source/Core/Core/HW/HSP/HSP_DeviceARAMExpansion.cpp:16]: These are also unused (presumably they were used by the Game Boy Player implementation?)

File: Source/Core/Core/HW/HSP/HSP_DeviceNull.cpp
Comment (Pokechu22) [inline Source/Core/Core/HW/HSP/HSP_DeviceNull.cpp:12]: None of these includes are used.

File: Source/Core/Core/HW/HSP/HSP_Device.h
Comment (Pokechu22) [inline Source/Core/Core/HW/HSP/HSP_Device.h:17]: Since this is an enum class and thus these values don't leak into the parent `HSP` namespace, they don't need to be prefixed with `HSPDEVICE_` anymore (and could be changed from all capse, so `None` and `ARAMExpansion`).

**Review by Pokechu22 — COMMENTED** [review]
Things look generally good enough.  I've got some minor include nitpicks and one naming change, but other than those everything looks good to me.

**Comment by Pokechu22** [comment]
@dolphin-emu-bot rebuild

**Review by Pokechu22 — APPROVED** [review]

**Comment by JMC47** [comment]
@dolphin-emu-bot rebuild

**Comment by JMC47** [comment]
I don't want this to bitrot again, and it's got an approval.  Good enough for me.

**Pull request merged**
